Alternative fix for initial size in wxGTK

vadz opened the pull request:
I'm not sure how much simpler this is than #2321 but it does have one advantage in my testing: it sets the min size of a window on which `SetSize()` is called correctly, i.e. it still can't be shrunk beyond its minimum size. And it also fixes a problem with `SetMinSize()`, even if it's less important, of course.

Please let me know which PR you prefer, but I'd like to merge (exactly) one of them very soon. TIA!

BTW, here is the example I used for testing:
```cpp
#include <wx/wx.h>

class App : public wxApp
{
    void NewFrame(bool big)
    {
      wxFrame* frame = new wxFrame (NULL, wxID_ANY, big ? "Big frame" : "Fitting frame");

      wxStaticText* label = new wxStaticText(frame, wxID_ANY, "");
      label->SetLabel(wxGetLibraryVersionInfo().GetVersionString());
      label->SetFont(label->GetFont().Scale(5));

      wxBoxSizer* sizer = new wxBoxSizer (wxVERTICAL);
      sizer->Add(label, wxSizerFlags().Expand());
      frame->SetSizerAndFit(sizer);

      if (big)
          frame->SetSize(wxSize(900,600));

      frame->Show(true);
    }

public:
    bool OnInit() override
    {
        if ( !wxApp::OnInit() )
            return false;

        NewFrame(false);
        NewFrame(true);

        return true;
    }
};

wxIMPLEMENT_APP(App);
```

vadz committed:
Factor out wxTopLevelWindowGTK::GTKDoAfterShow()

Just extract the code generating wxEVT_SHOW for TLWs in wxGTK in its own
function before modifying it to avoid having to do it in two places.

No real changes, this is a pure refactoring.

vadz committed:
Move workaround for initial TLW size to wxGTK itself

Add wxWindow::WXSetInitialFittingClientSize() instead of handling wxGTK
TLWs specially in the common wxSizer code and override it in wxGTK to
remember that we need to reset the client size once the window is shown.

This commit shouldn't result in any changes in the observed behaviour.

vadz committed:
Don't overwrite explicitly set size with pending client size

Calls to SetSize() were ignored in wxGTK if they happened after calling
wxSizer::Fit() since the changes of f655a52fba (Allow wxSizer::Fit() to
work properly when called from TLW ctor on GTK3, 2020-04-20) because we
overwrote the explicitly set size with the client size computed earlier
by the sizer.

Don't do this by explicitly forgetting the client size we were going to
set when SetSize() is called.

Closes #19134.

vadz committed:
Also make SetMinSize() after wxSizer::Fit() work correctly

This fixes a problem similar to that fixed in the previous commit but
for SetMinSize(), which was also ignored if done after calling
wxSizer::Fit() and before showing the window, as the explicitly set min
size was also overwritten by the pending min size computed from the
client size corresponding to the sizer fitting size.

The fix is similar too: just invalidate the pending minimum size if
SetMinSize() is called.

paulcor commented on include/wx/window.h:
```diff
@@ -1541,6 +1548,15 @@ class WXDLLIMPEXP_CORE wxWindowBase : public wxEvtHandler
 
+    // This internal function needs to be called to set the fitting client size
+    // (i.e. the minimum size determined by the window sizer) when the window
+    // decorations may yet be unknown, as is the case immediately after
```
I don't think this issue is related to window decorations. It's caused by GTK not updating cached style information (specifically font sizes, which affect best size calculations) until the TLW is shown.

paulcor commented on include/wx/window.h:
```diff
@@ -1541,6 +1548,15 @@ class WXDLLIMPEXP_CORE wxWindowBase : public wxEvtHandler
+    //
+    // It currently only does something different from the default
+    // implementation for TLWs in wxGTK.
+    virtual void WXSetInitialFittingClientSize(int flags);
```
Is `WX` prefix on member functions a convention I'm not aware of?

vadz replied:
Yes, I use `MSW`, `GTK` or `OSX` for port-specific private functions and `WX` for common ones. This is used for the existing `WXAdjustFontToOwnPPI()` and `WXSendContextMenuEvent()` in the same header you can run something like `git grep '\<WX[A-Z][a-z][A-Za-z]+\(' include` to see more of them.

vadz replied:
P.S. I did write about this to wx-dev a couple (few? many?) years ago, but it's hard to find the message in the archives and we probably should document this in some more visible place, but I don't know where exactly. Any ideas would be welcome.

paulcor commented on src/gtk/toplevel.cpp:
```diff
@@ -1163,6 +1165,12 @@ bool wxTopLevelWindowGTK::Show( bool show )
+    if (change && show)
+    {
+        // We may need to redo it after showing the window.
+        GTKUpdateClientSizeIfNecessary();
```
I think this is redundant. Won't the `map-event` handler have already done this?

vadz replied:
I don't quite understand when is `GTKDoAfterShow()` called from `gtk_frame_map_callback()`, i.e. how to reproduce [#9909](https://trac.wxwidgets.org/ticket/9909) referenced there, but in my test this doesn't happen, i.e. the window is always already considered to be shown. I suspect this might not happen at all any longer, but I'm definitely not 100% sure about this... And calling `GTKUpdateClientSizeIfNecessary()` twice, which is the worst that can happen here, should do no harm, AFAICS.

paulcor commented on src/gtk/toplevel.cpp:
```diff
@@ -1523,6 +1543,52 @@ wxTopLevelWindowGTK::DecorSize& wxTopLevelWindowGTK::GetCachedDecorSize()
+#ifdef __WXGTK3__
+    // Set the client size again if necessary, we should be able to do it
+    // correctly by now as the style cache should be up to date.
+    GTKUpdateClientSizeIfNecessary();
```
It might be better to do this before generating the `wxShowEvent`.

vadz replied:
Good point, thanks, changed.

paulcor commented on src/gtk/toplevel.cpp:
```diff
@@ -1523,6 +1543,52 @@ wxTopLevelWindowGTK::DecorSize& wxTopLevelWindowGTK::GetCachedDecorSize()
+
+    // But if we're not shown yet, also redo it later if the decoration size
+    // may change.
+    if ( !IsShown() && m_updateDecorSize )
```
Again, I don't think decoration size is involved, so the test for `m_updateDecorSize` seems wrong.

paulcor reviewed (commented):
I think I like your approach better. But I can't spend any more time on this until at least this evening (PDT).

vadz committed:
Update TLW client size in wxGTK before sending wxEVT_SHOW

If we have to change the client size, it's better to do this before
sending the event whose handler might rely on the size being already
correct.

Co-Authored-By: Paul Cornett <paulcor@users.noreply.github.com>

vadz commented:
NB: CI failure is due to [unrelated bug](https://trac.wxwidgets.org/ticket/19141)

If there are no objections, I'm going to merge this and make 3.1.5 source archives soon.
